Some tests on Rockchip RK3328 Nougat.

Mpeg2/4 and interlaced SD content already hardware decoded with MediaCodec and no problems, correctly deinterlaced, no need for software decoding.

Tests with
kodi-20190119-5936ad20-glTexSubImage2D-armeabi-v7a & 576i50_mpeg2_samples.ts
MediaCodec/MediaCodec Surface enabled: 20-30% CPU usage, perfectly hardware decoded and deinterlaced.
Auto frame-rate switching takes place twice, first to 25fps then a second later to 50hz.

MediaCodec/MediaCodec Surface disabled, Deinterlace half: 70-80% CPU usage for 1st core, 50% usage for other 3 cores. Video is watchable but with a bit more background noise than with MediaCodec enabled.

Kodi nightly builds - MediaCodec/MediaCodec Surface disabled, Deinterlace Half: 90% CPU usage for 1st core, 60% usage for other 3 cores, video less smooth.

So with glTexSubImage2D changes about a 10% average reduction in CPU usage with software decoding on Android for me with the test video.

------------------------
I want to try these changes on LibreELEC RK3328 but I get the following compile error:

`../xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp: In member function 'void CLinuxRendererGLES::LoadPlane(CLinuxRendererGLES::CYuvPlane&, int, unsigned int, unsigned int, int, int, void*)':
../xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp:279:21: error: 'GL_UNPACK_ROW_LENGTH' was not declared in this scope
       glGetIntegerv(GL_UNPACK_ROW_LENGTH, &pixelStore);
                     ^~~~~~~~~~~~~~~~~~~~
../xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp:279:21: note: suggested alternative: 'GL_UNPACK_ROW_LENGTH_EXT'
       glGetIntegerv(GL_UNPACK_ROW_LENGTH, &pixelStore);
                     ^~~~~~~~~~~~~~~~~~~~
                     GL_UNPACK_ROW_LENGTH_EXT
../xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp:286:19: error: 'GL_UNPACK_ROW_LENGTH' was not declared in this scope
     glPixelStorei(GL_UNPACK_ROW_LENGTH pixelStore);
                   ^~~~~~~~~~~~~~~~~~~~
../xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp:286:19: note: suggested alternative: 'GL_UNPACK_ROW_LENGTH_EXT'
     glPixelStorei(GL_UNPACK_ROW_LENGTH pixelStore);
                   ^~~~~~~~~~~~~~~~~~~~
                   GL_UNPACK_ROW_LENGTH_EXT
[1404/1524] Building CXX object build/platform/linux/CMakeFiles/linuxsupport.dir/DBusReserve.cpp.o
ninja: build stopped: subcommand failed.
make: *** [image] Error 1`

Here is the source code used for LibreELEC RK3328(Kodi part), also a Kodi 18 RC5.2 Leila branch.
https://github.com/Kwiboo/xbmc/tree/rockchip_18.0rc5.2-Leia/xbmc/cores/VideoPlayer/VideoRenderers

Using this I also just get a black screen with audio playing in LibreELEC RK3328.
https://github.com/fritsch/xbmc/commit/c527e095bbdd7539f7ca71c5286f7ca930287236

Since DRM content with inputstream can't be hardware decoded in Linux/LibreELEC, I want to test if any of these patches can improve software decoding for eg. H264 Netflix 720p which runs unstable with high CPU usage in LibreELEC.

--- a/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
@@ -264,6 +264,8 @@ void CLinuxRendererGLES::LoadPlane(CYuvPlane& plane, int type,
 
   // OpenGL ES does not support strided texture input.
   GLint pixelStore = -1;
+  unsigned int pixelStoreKey = -1;
+
   if (stride != static_cast<int>(width * bps))
   {
 #if HAS_GLES >= 3
@@ -274,6 +276,15 @@ void CLinuxRendererGLES::LoadPlane(CYuvPlane& plane, int type,
     {
       glGetIntegerv(GL_UNPACK_ROW_LENGTH, &pixelStore);
       glPixelStorei(GL_UNPACK_ROW_LENGTH, stride);
+      pixelStoreKey = GL_UNPACK_ROW_LENGTH;
+    }
+    else
+#elif defined (GL_UNPACK_ROW_LENGTH_EXT)
+    if (m_pRenderSystem->IsExtSupported("GL_EXT_unpack_subimage"))
+    {
+      glGetIntegerv(GL_UNPACK_ROW_LENGTH_EXT, &pixelStore);
+      glPixelStorei(GL_UNPACK_ROW_LENGTH_EXT, stride);
+      pixelStoreKey = GL_UNPACK_ROW_LENGTH_EXT;
     }
     else
 #endif
@@ -290,7 +301,7 @@ void CLinuxRendererGLES::LoadPlane(CYuvPlane& plane, int type,
   glTexSubImage2D(m_textureTarget, 0, 0, 0, width, height, type, GL_UNSIGNED_BYTE, pixelData);
 
   if (pixelStore >= 0)
-    glPixelStorei(GL_UNPACK_ROW_LENGTH, pixelStore);
+    glPixelStorei(pixelStoreKey, pixelStore);
 
   // check if we need to load any border pixels
   if (height < plane.texheight)

--- a/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
@@ -267,17 +267,25 @@ void CLinuxRendererGLES::LoadPlane(CYuvPlane& plane, int type,
   if (stride != static_cast<int>(width * bps))
   {
 #if HAS_GLES >= 3
-    glGetIntegerv(GL_UNPACK_ROW_LENGTH, &pixelStore);
-    glPixelStorei(GL_UNPACK_ROW_LENGTH, stride);
-#else
-    unsigned char *src(static_cast<unsigned char*>(data)),
-                  *dst(m_planeBuffer);
+    unsigned int verMajor, verMinor;
+    m_renderSystem->GetRenderVersion(verMajor, verMinor);
 
-    for (unsigned int y = 0; y < height; ++y, src += stride, dst += width * bpp)
-      memcpy(dst, src, width * bpp);
-
-    pixelData = m_planeBuffer;
+    if (verMajor >= 3)
+    {
+      glGetIntegerv(GL_UNPACK_ROW_LENGTH, &pixelStore);
+      glPixelStorei(GL_UNPACK_ROW_LENGTH, stride);
+    }
+    else
 #endif
+    {
+      unsigned char *src(static_cast<unsigned char*>(data)),
+                    *dst(m_planeBuffer);
+
+      for (unsigned int y = 0; y < height; ++y, src += stride, dst += width * bpp)
+        memcpy(dst, src, width * bpp);
+
+      pixelData = m_planeBuffer;
+    }
   }
   glTexSubImage2D(m_textureTarget, 0, 0, 0, width, height, type, GL_UNSIGNED_BYTE, pixelData);
 

--- a/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp
@@ -46,6 +46,9 @@ CLinuxRendererGLES::~CLinuxRendererGLES()
   UnInit();
 
   ReleaseShaders();
+
+  free(m_planeBuffer);
+  m_planeBuffer = nullptr;
 }
 
 CBaseRenderer* CLinuxRendererGLES::Create(CVideoBuffer *buffer)
@@ -260,18 +263,26 @@ void CLinuxRendererGLES::LoadPlane(CYuvPlane& plane, int type,
   glBindTexture(m_textureTarget, plane.id);
 
   // OpenGL ES does not support strided texture input.
+  GLint pixelStore = -1;
   if (stride != static_cast<int>(width * bps))
   {
-    unsigned char* src = static_cast<unsigned char*>(data);
-    for (unsigned int y = 0; y < height; ++y, src += stride)
-    {
-      glTexSubImage2D(m_textureTarget, 0, 0, y, width, 1, type, GL_UNSIGNED_BYTE, src);
-    }
+#if HAS_GLES >= 3
+    glGetIntegerv(GL_UNPACK_ROW_LENGTH, &pixelStore);
+    glPixelStorei(GL_UNPACK_ROW_LENGTH, stride);
+#else
+    unsigned char *src(static_cast<unsigned char*>(data)),
+                  *dst(m_planeBuffer);
+
+    for (unsigned int y = 0; y < height; ++y, src += stride, dst += width * bpp)
+      memcpy(dst, src, width * bpp);
+
+    pixelData = m_planeBuffer;
+#endif
   }
-  else
-  {
-    glTexSubImage2D(m_textureTarget, 0, 0, 0, width, height, type, GL_UNSIGNED_BYTE, pixelData);
-  }
+  glTexSubImage2D(m_textureTarget, 0, 0, 0, width, height, type, GL_UNSIGNED_BYTE, pixelData);
+
+  if (pixelStore >= 0)
+    glPixelStorei(GL_UNPACK_ROW_LENGTH, pixelStore);
 
   // check if we need to load any border pixels
   if (height < plane.texheight)
@@ -1261,6 +1272,8 @@ bool CLinuxRendererGLES::CreateYV12Texture(int index)
   im.planesize[0] = im.stride[0] * im.height;
   im.planesize[1] = im.stride[1] * (im.height >> im.cshift_y);
   im.planesize[2] = im.stride[2] * (im.height >> im.cshift_y);
+
+  m_planeBuffer = static_cast<unsigned char*>(realloc(m_planeBuffer, m_sourceHeight * m_sourceWidth * im.bpp));
 
   for (int i = 0; i < 3; i++)
   {

--- a/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.h
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.h
@@ -201,6 +201,7 @@ protected:
   bool m_fullRange;
   AVColorPrimaries m_srcPrimaries;
   bool m_toneMap = false;
+  unsigned char* m_planeBuffer = nullptr;
 
   // clear colour for "black" bars
   float m_clearColour{0.0f};

